src/tpm2.rs

```diff
@@ -0,0 +1,957 @@
+                        Some(handle) => {
+                            self.update_tpmdata("ek_handle", &handle)?
+                        }
+                    }
+                },
+            }
+
+            ret_f.get(tf_path).map_or_else(
+                || {
+                    Err(KeylimeTpmError::new_tpm_rust_error(
+                        "tpm2_readpublic fail, ek key is missing is output file.",
+                    ))
+                },
+                |ek_tpm| {
+                    self.update_tpmdata("ek_tpm", &base64::encode(&ek_tpm))
+                },
+            )?;
+            self.update_tpmdata("ek_pw", &ek_pw)?;
+            self.write_tpmdata()
+        })?;
+        Ok(())
+    }
+
+    fn get_pub_ek(&mut self) -> Result<(), KeylimeTpmError> {
+        let tf = NamedTempFile::new()?;
+        let tf_path = self.temp_file_get_path(&tf)?;
+
+        let cmd = match self.legacy_tools {
+            true => format!(
+                "tpm2_readpublic -H {} -o {} -f pem",
+                hex::encode(self.get_tpmdata("ek_handle")),
+                tf_path
+            ),
+            false => format!(
+                "tpm2_readpublic -c {} -o {} -f pem",
+                hex::encode(self.get_tpmdata("ek_handle")),
+                tf_path
+            ),
+        };
+
+        self.run(cmd, Some(vec![tf_path])).and_then(|(_, ret_f)| {
+            ret_f.get(tf_path).map_or_else(
+                || {
+                    Err(KeylimeTpmError::new_tpm_rust_error(
+                        "Error: failed to get public ek key.",
+                    ))
+                },
+                |ek| self.update_tpmdata("ek", ek),
+            )
+        })?;
+        Ok(())
+    }
+
+    fn get_pub_aik(&mut self) -> Result<(), KeylimeTpmError> {
+        if let false = self.legacy_tools {
+            return Err(KeylimeTpmError::new_tpm_rust_error(
+                "get public aik doesn't apply to modern tpm2-tools",
+            ));
+        }
+
+        let tf = NamedTempFile::new()?;
+        let tf_path = self.temp_file_get_path(&tf)?;
+
+        self.run(
+            format!(
+                "tpm2_readpublic -H {} -o {} -f pem ",
+                hex::encode(self.get_tpmdata("aik_handle")),
+                tf_path
+            ),
+            None,
+        )
+        .and_then(|(_, file_ret)| {
+            if file_ret.is_empty() {
+                return Err(KeylimeTpmError::new_tpm_rust_error(
+                    "unable to read public aik",
+                ));
+            }
+            self.update_tpmdata(
+                "aik",
+                file_ret.get(tf_path).ok_or_else(|| {
+                    KeylimeTpmError::new_tpm_rust_error(
+                        "Error: get public aik key failed.",
+                    )
+                })?,
+            )
+        })?;
+        Ok(())
+    }
+
+    pub fn get_tpm_version(&self) -> i32 {
```

Comment:
What is this and why is it a function instead of a constant?

Reply:
This is based on the implementation of the python version of Keylime. Looks like a constant is a good practice here. What do you think? @jetwhiz .

Reply:
Why would it matter what the python version has as constants versus functions?  This is a separate codebase.  If we're not using it as a function, it shouldn't be a function.